### PR TITLE
API - A bit more RESTful

### PR DESCRIPTION
**Pull request by gen2fish**
I tried to take a stab at https://github.com/grocy/grocy/issues/139

I left the generic entries as is since it seemed a lot of logic was built in and I didn't want to break anything big. I haven't spun up a Dev instance yet to test, but I would hate to have duplicate work.  If you see something glaringly bad, I can try to fix it. I'll do some testing in a docker lab here soon.

### CODE REVIEW

File: routes.php [hunk 1]
```diff
@@ -102,46 +102,46 @@
...
-	$this->get('/stock/clear-shopping-list', '\Grocy\Controllers\StockApiController:ClearShoppingList');
-	$this->get('/stock/external-barcode-lookup/{barcode}', '\Grocy\Controllers\StockApiController:ExternalBarcodeLookup');
-	$this->get('/stock/undo-booking/{bookingId}', '\Grocy\Controllers\StockApiController:UndoBooking');
+	$this->post('/stock/{productId}/add/{amount}', '\Grocy\Controllers\StockApiController:AddProduct');
+	$this->post('/stock/{productId}/consume/{amount}', '\Grocy\Controllers\StockApiController:ConsumeProduct');
+	$this->post('/stock/{productId}/open/{amount}', '\Grocy\Controllers\StockApiController:OpenProduct');
+	$this->post('/stock/{productId}/inventory/{newAmount}', '\Grocy\Controllers\StockApiController:InventoryProduct');
+	$this->get('/stock/{productId}', '\Grocy\Controllers\StockApiController:ProductDetails');
+	$this->get('/stock/{productId}/pricehistory', '\Grocy\Controllers\StockApiController:ProductPriceHistory');
+	$this->get('/stock/{productId}/entires', '\Grocy\Controllers\StockApiController:ProductStockEntries');
```
Comment (MrTarantula) [inline routes.php]: entires => entries

File: routes.php [hunk 2]
```diff
@@ -102,46 +102,46 @@
...
+	$this->get('/stock', '\Grocy\Controllers\StockApiController:CurrentStock');
+	$this->get('/stock/volatil', '\Grocy\Controllers\StockApiController:CurrentVolatilStock');
```
Comment (MrTarantula) [inline routes.php]: volatil => volatile

**Comment by MrTarantula** [comment]
I like how the endpoints are structured, Good job. This is just a nit, but can you group the routes by verb, so all GETs are first, then POSTs, etc?

**Comment by MrTarantula** [comment]
Also, there's the question of breaking API. Should it be a hard break? Or start versioning the API, e.g. `/api/v2/stock/{productId}`? @berrnd your call.

File: public/viewjs/batteriesoverview.js [hunk 3]
```diff
@@ -61,15 +61,15 @@ $(document).on('click', '.track-charge-cycle-button', function(e)
...
 
 	Grocy.FrontendHelpers.BeginUiBusy();
-	
+
 	var batteryId = $(e.currentTarget).attr('data-battery-id');
 	var batteryName = $(e.currentTarget).attr('data-battery-name');
 	var trackedTime = moment().format('YYYY-MM-DD HH:mm:ss');
 
-	Grocy.Api.Get('batteries/track-charge-cycle/' + batteryId + '?tracked_time=' + trackedTime,
+	Grocy.Api.Post('batteries/' + batteryId + '/charged?tracked_time=' + trackedTime,
```
Comment (MrTarantula) [inline public/viewjs/batteriesoverview.js]: Does this work? POST with query parameters is not typical. Data should be sent in POST body. POSTs will require changes on server side as well.

**Comment by berrnd** [comment]
I would prefer a hard break, so a major version 2.x when this is ready. Currently it would be no problem, because the logic behind the API will not really change for now, but when this happens, it's maybe too cumbersome to keep the old version of the API alive...

File: routes.php (hunk 1 above)
Comment (gen2fish) [reply routes.php]: Fixed in https://github.com/grocy/grocy/pull/140/commits/e416695cc4aa7c77b9f5e2d0d4f7dbbfef843856

File: routes.php (hunk 2 above)
Comment (gen2fish) [reply routes.php]: https://github.com/grocy/grocy/pull/140/commits/e416695cc4aa7c77b9f5e2d0d4f7dbbfef843856

**Comment by gen2fish** [comment]
Did some cleaning in https://github.com/grocy/grocy/pull/140/commits/e416695cc4aa7c77b9f5e2d0d4f7dbbfef843856

Moved the gets on top and posts on bottom per section and PUTs for user/object edits.

Working on removing the verbs in the routes.

**Comment by gen2fish** [comment]
Would you rather have the same route used for GET and POST or for new items have a /new?

```php
$this->get(/object/{entity} // Get Object
$this->post(/object/{entity} // Create Object
```

Or

```php
$this->get(/object/{entity} // Get Object
$this->post(/object/{entity}/new // Create Object
```

I would suspect we can use the same route with different methods for each section.

**Comment by berrnd** [comment]
I think the later (same route, different methods) is the recommended way...

**Comment by gen2fish** [comment]
Verb-less generic routes

https://github.com/grocy/grocy/pull/140/commits/6b41e5212c27c881b96379fb993a7bbfd6636ca0

I'll spin up a test instance when I got home tonight and see if everything is broken.

I'll move on the Stock with the amount in the URL string next and try to put it in the JSON body.

**Comment by berrnd** [comment]
Thanks.

At least all `PUT` routes will be affected, because `Grocy.Api.Put` does not exist as a JS function yet.
Also [grocy.openapi.json](https://github.com/grocy/grocy/blob/master/grocy.openapi.json) needs to be adjusted to reflect the changes in the integrated [Swagger UI](https://demo-en.grocy.info/api) instance.

**Comment by gen2fish** [comment]
I created the function quick in https://github.com/grocy/grocy/pull/140/commits/3db750e2119bb546a106c85456c030a14d23bd6f

**Comment by gen2fish** [comment]
Spun up a test instance. Testing is going well.  Only one issue with the volatile stock route.  Getting this error on startup

```
Fatal error: Uncaught FastRoute\BadRouteException: Static route "/api/stock/volatile" is shadowed by previously defined variable route "/api/stock/([^/]+)" for method "GET" in /www/vendor/nikic/fast-route/src/DataGenerator/RegexBasedAbstract.php:95 Stack trace: #0 /www/vendor/nikic/fast-route/src/DataGenerator/RegexBasedAbstract.php(30): FastRoute\DataGenerator\RegexBasedAbstract->addStaticRoute('GET', Array, 'route74') #1 /www/vendor/nikic/fast-route/src/RouteCollector.php(44): FastRoute\DataGenerator\RegexBasedAbstract->addRoute('GET', Array, 'route74') #2 /www/vendor/slim/slim/Slim/Router.php(227): FastRoute\RouteCollector->addRoute(Array, '/api/stock/vola...', 'route74') #3 /www/vendor/nikic/fast-route/src/functions.php(25): Slim\Router->Slim\{closure}(Object(FastRoute\RouteCollector)) #4 /www/vendor/slim/slim/Slim/Router.php(238): FastRoute\simpleDispatcher(Object(Closure), Array) #5 /www/vendor/slim/slim/Slim/Router.php(191): Slim\Router->createDispatcher() #6 /www/vendor/slim/slim/Slim/App.php(585): Slim\Router->di in /www/vendor/nikic/fast-route/src/DataGenerator/RegexBasedAbstract.php on line 95
```

Once I comment out line 116 in routes.php
```php
	$this->get('/stock/volatile', '\Grocy\Controllers\StockApiController:CurrentVolatilStock');
```

Application starts successfully.

So far I have tested the Stock GETs

**Comment by MrTarantula** [comment]
Do the new POSTs work? They shouldn't, as far as I can tell. Although I'm not really a PHP dev, I'd like to contribute something. I am a web dev and MVC is similar enough in any language.

I should have a bit of time this weekend to work on the POSTs on the server side, unless you get to it first. I'll submit a PR to your branch when I'm ready.

**Comment by gen2fish** [comment]
I haven't gotten to modifying the posts to have the data in the json body instead of the url. They work in their current form but I haven't spent much time on getting them fixed. Was hoping to get some time this weekend myself.

**Comment by berrnd** [comment]
I'll merge this now and complete it this weekend - thanks @gen2fish and @MrTarantula! :)

Further progress will be tracked in #139.

**Pull request merged**